batnode.js

```diff
@@ -250,25 +251,59 @@ class BatNode {
       this.getHostNode(currentCopy, afterHostNodeIsFound)
     }
   }
+  combineShardsAfterWaitTime(waitTime, fileName, distinctShards) {
+    return new Promise(resolve => {
+      setTimeout(() => resolve(fileUtils.assembleShards(fileName, distinctShards)), waitTime);
+    });
+  }
+
+  async asyncCallAssembleShards(waitTime, fileName, distinctShards) {
+    try {
+      console.log("waiting time in ms: ", waitTime);
+      const result = await this.combineShardsAfterWaitTime(waitTime, fileName, distinctShards);
+      return result;
+    } catch (e) {
+      console.error(e);
+    }
+  }
 
-  issueRetrieveShardRequest(shardId, hostBatNode, options, finishCallback){
+  issueRetrieveShardRequest(shardId, hostBatNode, manifest, options, finishCallback){
    let { saveShardAs, distinctIdx, distinctShards, fileName } = options
+
    let client = this.connect(hostBatNode.port, hostBatNode.host, () => {
     let message = {
       messageType: 'RETRIEVE_FILE',
       fileName: shardId
     }
 
-    client.on('data', (data) => {
-      fs.writeFileSync(`./shards/${saveShardAs}`, data, 'utf8')
+    if (!fs.existsSync('./shards/')){ fs.mkdirSync('./shards/'); }
+
+    const fileDestination = './shards/' + saveShardAs;
+    let shardStream = fs.createWriteStream(fileDestination);
+
+    const completeFileSize = manifest.fileSize;
+    const waitTime = Math.floor(completeFileSize/16000);  // set the divided amount slightly below 16kb ~ 16384 (the default high watermark for read/write streams)
+
+    process.setMaxListeners(100);
```

Comment:
This is interesting. Could you leave a comment here explaining what this is / does and why it's here?

Reply:
This is a line to prevent Possible EventEmitter memory leak detected-will add it as comment

Reply:
Correction: Possible EventEmitter memory leak detected actually happens when the client still trying to connect with other nodes while other nodes have connection problem. This shouldn't be in this particular method. Will move it to `nodeCLIConnectionCallback`